modules/SerialCommHub/tiny_modbus_rtu.cpp

```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 Pionix GmbH and Contributors to EVerest
+
+// TODOs:
+// - sometimes we receive 0 bytes from sofar, find out why
+// - implement echo removal for chargebyte
+// - implement GPIO to switch rx/tx
+
+#include "tiny_modbus_rtu.hpp"
+
+#include "crc16.hpp"
+#include <errno.h>
+#include <fcntl.h>
+#include <fmt/core.h>
+#include <iostream>
+#include <string.h>
+#include <string>
+#include <unistd.h>
+
+namespace tinymod {
+
+static void hexdump(const char* text, uint8_t* msg, int msg_len) {
+    std::cout << text;
+    for (int i = 0; i < msg_len; i++) {
+        std::cout << std::hex << (int)msg[i] << " ";
+    }
+    std::cout << std::endl;
+}
+
+static void set_cksum(uint8_t* msg, int msg_len) {
+    if (msg_len < 5)
+        return;
+    uint16_t crc_sum = MODBUS_CRC16(msg, msg_len - 2);
+    memcpy(msg + msg_len - 2, &crc_sum, 2);
+}
+
+static bool validate_cksum(const uint8_t* msg, int msg_len) {
+    if (msg_len < 5)
+        return false;
+    // check crc
+    uint16_t crc_sum = MODBUS_CRC16(msg, msg_len - 2);
+    uint16_t crc_msg;
+    memcpy(&crc_msg, msg + msg_len - 2, 2);
+    return (crc_msg == crc_sum);
+}
+
+static std::vector<uint16_t> decode_reply(const uint8_t* buf, int len, uint8_t device_address, function_code function) {
+    std::vector<uint16_t> result;
+    if (len < 5) {
+        EVLOG_error << fmt::format("Packet too small: {} bytes.", len);
+        return result;
+    }
+    if (device_address != buf[device_address_pos]) {
+        EVLOG_error << fmt::format("Device address mismatch: expected: {} received: {}", device_address,
+                                   buf[device_address_pos]);
+        return result;
+    }
+
+    bool exception = false;
+    uint8_t function_code_recvd = buf[function_code_pos];
+    if (function_code_recvd & (1 << 7)) {
+        // highest bit is set for exception reply
+        exception = true;
+        // clear error bit
+        function_code_recvd &= (1 << 7);
+    }
```

Comment:
this could also be a helper function like
```
auto check_for_exception(uint8_t received_function_code) {
  return function_code_recvd & (1 << 7);
}
```
or you just do `bool exception = function_code_recvd & (1 << 7)` and later then `if(function != (function_code_recvd & (1 << 7))` of course using defines for `(1 << 7)`

I don't get here, why `function_code_recvd &= (1 << 7)` clears an error bit? Doesn't it only leave it?

Reply:
Yes, it does not clear the error but it should. I'll add the missing ~